lightning/src/ln/peers/handshake/states.rs

```diff
@@ -0,0 +1,906 @@
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+use bitcoin::secp256k1;
+
+use bitcoin::hashes::{Hash, HashEngine};
+use bitcoin::hashes::sha256::Hash as Sha256;
+use bitcoin::secp256k1::{SecretKey, PublicKey};
+
+use ln::peers::{chacha, hkdf5869rfc};
+use ln::peers::conduit::{Conduit, SymmetricKey};
+use ln::peers::handshake::acts::{Act, ActBuilder, ACT_ONE_LENGTH, ACT_TWO_LENGTH, ACT_THREE_LENGTH, EMPTY_ACT_ONE, EMPTY_ACT_TWO, EMPTY_ACT_THREE};
+
+// Alias type to help differentiate between temporary key and chaining key when passing bytes around
+type ChainingKey = [u8; 32];
+
+// Generate a SHA-256 hash from one or more elements concatenated together
+macro_rules! concat_then_sha256 {
+	( $( $x:expr ),+ ) => {{
+		let mut sha = Sha256::engine();
+		$(
+			sha.input($x.as_ref());
+		)+
+		Sha256::from_engine(sha)
+	}}
+}
+
+pub(super) enum HandshakeState {
+	InitiatorStarting(InitiatorStartingState),
+	ResponderAwaitingActOne(ResponderAwaitingActOneState),
+	InitiatorAwaitingActTwo(InitiatorAwaitingActTwoState),
+	ResponderAwaitingActThree(ResponderAwaitingActThreeState),
+	Complete(Option<(Conduit, PublicKey)>),
+}
+
+// Trait for all individual states to implement that ensure HandshakeState::next() can
+// delegate to a common function signature. May transition to the same state in the event there are
+// not yet enough bytes to move forward with the handshake.
+pub(super) trait IHandshakeState {
+	fn next(self, input: &[u8]) -> Result<(Option<Act>, HandshakeState), String>;
+}
+
+// Enum dispatch for state machine. Single public interface can statically dispatch to all states
+impl HandshakeState {
+	pub(super) fn new_initiator(initiator_static_private_key: &SecretKey, responder_static_public_key: &PublicKey, initiator_ephemeral_private_key: &SecretKey) -> Self {
+		HandshakeState::InitiatorStarting(InitiatorStartingState::new(initiator_static_private_key.clone(), initiator_ephemeral_private_key.clone(), responder_static_public_key.clone()))
+	}
+	pub(super) fn new_responder(responder_static_private_key: &SecretKey, responder_ephemeral_private_key: &SecretKey) -> Self {
+		HandshakeState::ResponderAwaitingActOne(ResponderAwaitingActOneState::new(responder_static_private_key.clone(), responder_ephemeral_private_key.clone()))
+	}
+}
+
+impl IHandshakeState for HandshakeState {
+	fn next(self, input: &[u8]) -> Result<(Option<Act>, HandshakeState), String> {
+		match self {
+			HandshakeState::InitiatorStarting(state) => { state.next(input) },
+			HandshakeState::ResponderAwaitingActOne(state) => { state.next(input) },
+			HandshakeState::InitiatorAwaitingActTwo(state) => { state.next(input) },
+			HandshakeState::ResponderAwaitingActThree(state) => { state.next(input) },
+			HandshakeState::Complete(_conduit) => { panic!("no acts left to process") }
+		}
+	}
+}
+
+// Handshake state of the Initiator prior to generating Act 1
+pub(super) struct InitiatorStartingState {
```

Comment:
Do the structs here need to be pub(super)? It looks like they're only ever accessed in two tests to check state, which could just as easily be a test-only method returning a state name or something.

Reply:
Yes, they are contained by the HandshakeState enums which can't leak private types in the mod.rs usages.

Reply:
Right, seems like we could make `HandshakeState` a struct and move the enum inside it to reduce the exposed surface of states.rs significantly?

Reply:
What is the point of reducing the exposed surface? This is just an enum used inside the handshake module. Would you rather just inline it all in one file? That is sort of what it sounds like.

This is how the enum dispatch pattern works in Rust and it seems like we are just talking past each other. If you are uncomfortable with this pattern in your codebase feel free to have @arik-so or @ariard change it or discard the PRs.